File: Code.java
// Definition for singly-linked list:
// class ListNode<T> {
//   ListNode(T x) {
//     value = x;
//   }
//   T value;
//   ListNode<T> next;
// }
//
ListNode<Integer> addTwoHugeNumbers(ListNode<Integer> a, ListNode<Integer> b) {
    int remainder = 0, sum = 0;
    ListNode<Integer> node = null;
    ListNode<Integer> temp = null;
    ListNode<Integer> head = null;


    a = reverse(a);
    b = reverse(b);
    
    while(a != null || b != null)
    {       
        sum = (a == null ? 0 : a.value) + (b == null ? 0 : b.value);
        
        if(remainder > 0)
        {
            sum += remainder;
            remainder = 0;   
        }
        
        if(sum > 9999)
        {
            remainder = 1;
            sum -= 10000;
        }
        
        temp = new ListNode<Integer>(sum);
        
        if(node == null)
        {    
            node = temp;
        }
        else{
            head.next = temp;
        }
        
        head = temp;
        if(a != null)
            a = a.next;
        if(b != null)
            b = b.next;
    }
    
    if(remainder == 1)
    {
        head.next = new ListNode<Integer>(1);
    }
    
    node = reverse(node);

    
    return node;
}

ListNode reverse(ListNode<Integer> a)
{
    ListNode<Integer> curr = a;
    ListNode<Integer> next = null;
    ListNode<Integer> prev = null;
    
    while(curr != null)
    {
        next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }
    
    a = prev;
    return a;
}

File: Code.c
//
// Definition for singly-linked list:
// typedef struct list_##name {
//   type value;
//   struct list_##name *next;
// } list_##name;
//
// list_##name *alloc_list_##name(type v) {
//   list_##name *l = calloc(1, sizeof(list_##name));
//   l->value = v;
//   return l;
// }
//
//


void reverse(list_integer ** l){
    list_integer * next;
    list_integer * prev = NULL;
    list_integer * curr = *l;
    
    while(curr != NULL){
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    
    *l = prev;
    
}

void print_linked_list(list_integer * list){
    while(list){
        printf("%d ", list->value);
        list = list->next;
    }
}

list_integer * addTwoHugeNumbers(list_integer * a, list_integer * b) {
    reverse(&a);
    reverse(&b);
    //print_linked_list(a);
    
    
    list_integer * head = NULL;
    int rem = 0;
    
    while(a != NULL || b != NULL){
        int v_a = 0;
        int v_b = 0;
        if(a != NULL)
            v_a = a->value; //printf(" %d h", a->value);}
        if(b != NULL)
            v_b = b->value; //printf(" %d t", b->value);}
        
        int total = v_a + v_b + rem;
        rem = 0;
        
        if(total > 9999){
            total -= 10000;
            rem = 1;
        }
        
        list_integer * res;
        res = alloc_list_integer(total);
        res->next = head;
        head = res;
   
        if(a != NULL) a = a->next;
        if(b != NULL) b = b->next;
    }
    
    if(rem > 0){
        list_integer * res;
        res = alloc_list_integer(1);
        res->next = head;
        head = res;
    }
    
    return head;
}
